### PR TITLE
feat(precompile): use `PolarContext` for Precompile calls

### PR DESCRIPTION
**Pull request by shampoobera**

### CODE REVIEW

File: cosmos/lib/evm.go [hunk 1]
```diff
@@ -33,15 +33,15 @@ import (
 	"pkg.berachain.dev/polaris/lib/utils"
 )
 
-// TODO: Add these functions to the ethprecompile.EVM object itself to allow enforcing calls into
+// TODO: Add these functions to thevm.PrecompileEVM object itself to allow enforcing calls into
```
Comment (calbera) [inline cosmos/lib/evm.go:36]:
```suggestion
// TODO: Add these functions to the EVM object itself to allow enforcing calls into
```

**Comment by itsdevbear** [comment]
yeah @shampoobera this is straight fire my g

File: cosmos/lib/bank.go [hunk 2]
```diff
@@ -45,7 +45,6 @@ func MintCoinsToAddress(
 	if err := bk.MintCoins(ctx, moduleAcc, coins); err != nil {
 		return err
 	}
-
```
Comment (itsdevbear) [inline cosmos/lib/bank.go:48]: re-add the line

**Comment by itsdevbear** [comment]
@shampoobera that 1 deleted line then good to merge

**Review by itsdevbear — APPROVED** [review]

File: cosmos/precompile/erc20/erc20.go [hunk 3]
```diff
@@ -129,84 +124,113 @@ func (c *Contract) Erc20AddressForCoinDenom(
...
 	denom string,
 	amount *big.Int,
 ) ([]any, error) {
-	err := c.transferCoinToERC20(ctx, evm, value, denom, caller, caller, amount)
+	err := c.transferCoinToERC20(ctx,
+		vm.UnwrapPolarContext(ctx).Evm(),
+		vm.UnwrapPolarContext(ctx).MsgValue(),
+		denom,
+		vm.UnwrapPolarContext(ctx).MsgSender(),
+		vm.UnwrapPolarContext(ctx).MsgSender(),
```
Comment (calbera) [inline cosmos/precompile/erc20/erc20.go:135]: do vm.UnwrapPolarContext(ctx) once

File: cosmos/precompile/erc20/erc20.go [hunk 4]
```diff
@@ -129,84 +124,113 @@ func (c *Contract) Erc20AddressForCoinDenom(
...
 	denom string,
 	owner common.Address,
 	recipient common.Address,
 	amount *big.Int,
 ) ([]any, error) {
-	err := c.transferCoinToERC20(ctx, evm, value, denom, owner, recipient, amount)
+	err := c.transferCoinToERC20(
+		ctx,
+		vm.UnwrapPolarContext(ctx).Evm(),
+		vm.UnwrapPolarContext(ctx).MsgValue(),
```
Comment (calbera) [inline cosmos/precompile/erc20/erc20.go:152]: +1

File: cosmos/precompile/erc20/erc20.go [hunk 5]
```diff
@@ -129,84 +124,113 @@ func (c *Contract) Erc20AddressForCoinDenom(
...
 	recipient common.Address,
 	amount *big.Int,
 ) ([]any, error) {
-	err := c.transferCoinToERC20(ctx, evm, value, denom, caller, recipient, amount)
+	err := c.transferCoinToERC20(
+		ctx,
+		vm.UnwrapPolarContext(ctx).Evm(),
+		vm.UnwrapPolarContext(ctx).MsgValue(),
+		denom,
+		vm.UnwrapPolarContext(ctx).MsgSender(),
```
Comment (calbera) [inline cosmos/precompile/erc20/erc20.go:173]: +1

File: cosmos/precompile/erc20/erc20.go [hunk 6]
```diff
@@ -129,84 +124,113 @@ func (c *Contract) Erc20AddressForCoinDenom(
...
 	amount *big.Int,
 ) ([]any, error) {
-	err := c.transferERC20ToCoin(ctx, caller, evm, token, caller, caller, amount)
+	err := c.transferERC20ToCoin(
+		ctx,
+		vm.UnwrapPolarContext(ctx).MsgSender(),
+		vm.UnwrapPolarContext(ctx).Evm(),
+		token,
+		vm.UnwrapPolarContext(ctx).MsgSender(),
+		vm.UnwrapPolarContext(ctx).MsgSender(),
```
Comment (calbera) [inline cosmos/precompile/erc20/erc20.go:192]: +1

File: cosmos/precompile/erc20/erc20.go [hunk 7]
```diff
@@ -129,84 +124,113 @@ func (c *Contract) Erc20AddressForCoinDenom(
...
 	token common.Address,
 	owner common.Address,
 	recipient common.Address,
 	amount *big.Int,
 ) ([]any, error) {
-	err := c.transferERC20ToCoin(ctx, caller, evm, token, owner, recipient, amount)
+	err := c.transferERC20ToCoin(
+		ctx,
+		vm.UnwrapPolarContext(ctx).MsgSender(),
+		vm.UnwrapPolarContext(ctx).Evm(),
```
Comment (calbera) [inline cosmos/precompile/erc20/erc20.go:209]: +1

File: cosmos/precompile/erc20/erc20.go [hunk 8]
```diff
@@ -129,84 +124,113 @@ func (c *Contract) Erc20AddressForCoinDenom(
...
 	recipient common.Address,
 	amount *big.Int,
 ) ([]any, error) {
-	err := c.transferERC20ToCoin(ctx, caller, evm, token, caller, recipient, amount)
+	err := c.transferERC20ToCoin(
+		ctx,
+		vm.UnwrapPolarContext(ctx).MsgSender(),
+		vm.UnwrapPolarContext(ctx).Evm(),
+		token,
+		vm.UnwrapPolarContext(ctx).MsgSender(),
```
Comment (calbera) [inline cosmos/precompile/erc20/erc20.go:230]: +1

File: eth/core/precompile/method.go [hunk 9]
```diff
@@ -73,7 +74,9 @@ func NewMethod(
 }
 
 // Call executes the precompile's executable with the given context and input arguments.
-func (m *Method) Call(ctx []reflect.Value, input []byte) ([]byte, error) {
+//
+//nolint:revive // needed for reflection.
+func (m *Method) Call(sc Registrable, ctx context.Context, input []byte) ([]byte, error) {
```
Comment (calbera) [inline eth/core/precompile/method.go:79]: make sc of type `StatefulImpl`

File: eth/core/vm/context.go [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+
+func (c *PolarContext) Done() <-chan struct{} {
+	return c.baseCtx.Done()
+}
+
+func (c *PolarContext) Err() error {
+	return c.baseCtx.Err()
+}
+
+func (c *PolarContext) Value(key any) any {
```
Comment (calbera) [inline eth/core/vm/context.go:105]: also add the `WithValue` func

File: eth/core/vm/context.go [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+// PolarContext. It panics if a Context was not properly attached.
+func UnwrapPolarContext(ctx context.Context) *PolarContext {
+	if polarCtx, ok := utils.GetAs[*PolarContext](ctx); ok {
+		return polarCtx
+	}
+	return utils.MustGetAs[*PolarContext](ctx.Value(PolarContextKey))
+}
+
+// WithVaue attaches a value to the context.
+func (c *PolarContext) WithValue(key, value interface{}) *PolarContext {
```
Comment (calbera) [inline eth/core/vm/context.go:123]:
```suggestion
func (c *PolarContext) WithValue(key, value any) *PolarContext {
```

File: eth/core/vm/context.go [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+
+// Context is the context for a Polaris EVM execution.
+type PolarContext struct {
+	baseCtx   context.Context
+	evm       PrecompileEVM
+	msgSender common.Address
+	msgValue  *big.Int
+}
+
+// NewPolarContext creates a new polar.Context given an EVM call request.
```
Comment (calbera) [inline eth/core/vm/context.go:49]:
```suggestion
// NewPolarContext creates a new PolarContext given an EVM call request.
```

File: eth/core/vm/context.go [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+	"time"
+
+	"pkg.berachain.dev/polaris/eth/common"
+	"pkg.berachain.dev/polaris/lib/utils"
+)
+
+// ContextKey defines a type alias for a stdlib Context key.
+type ContextKey string
+
+// PolarContextKey is the key in the context.Context which holds the polar.Context.
```
Comment (calbera) [inline eth/core/vm/context.go:35]:
```suggestion
// PolarContextKey is the key in the context.Context which holds the PolarContext.
```

File: eth/core/vm/context.go [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+const PolarContextKey ContextKey = "polar-context"
+
+// Compile-time assertion that polar.Context implements context.Context.
```
Comment (calbera) [inline eth/core/vm/context.go:38]:
```suggestion
// Compile-time assertion that PolarContext implements context.Context.
```

File: eth/core/vm/context.go [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+func (c *PolarContext) Block() *BlockContext {
+	return c.evm.GetContext()
+}
+
+// =============================================================================
+// context.Context implementation
+// =============================================================================
+
+//nolint:nonamedreturns // context package uses named returns.
+func (c *PolarContext) Deadline() (deadline time.Time, ok bool) {
```
Comment (calbera) [inline eth/core/vm/context.go:93]:
```suggestion
func (c *PolarContext) Deadline() (time.Time, bool) {
```

File: eth/core/vm/context.go [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+		return polarCtx
+	}
+	return utils.MustGetAs[*PolarContext](ctx.Value(PolarContextKey))
+}
+
+// WithVaue attaches a value to the context.
+func (c *PolarContext) WithValue(key, value any) *PolarContext {
+	c.baseCtx = context.WithValue(c.baseCtx, key, value)
+	return c
+}
```
Comment (calbera) [inline eth/core/vm/context.go:125]: move this under the `Value` function

**Review by calbera — APPROVED** [review]

File: eth/core/precompile/method.go [hunk 17]
```diff
@@ -73,7 +74,9 @@ func NewMethod(
 }
 
 // Call executes the precompile's executable with the given context and input arguments.
-func (m *Method) Call(ctx []reflect.Value, input []byte) ([]byte, error) {
+//
+//nolint:revive // needed for reflection.
+func (m *Method) Call(sc StatefulImpl, ctx context.Context, input []byte) ([]byte, error) {
```
Comment (calbera) [inline eth/core/precompile/method.go:79]:
```suggestion
func (m *Method) Call(si StatefulImpl, ctx context.Context, input []byte) ([]byte, error) {
```

File: eth/core/precompile/method.go [hunk 18]
```diff
@@ -87,7 +90,11 @@ func (m *Method) Call(ctx []reflect.Value, input []byte) ([]byte, error) {
 	}
 
 	// Call the executable
-	results := m.execute.Call(append(ctx, reflectedUnpackedArgs...))
+	results := m.execute.Call(append(
+		[]reflect.Value{
+			reflect.ValueOf(sc),
```
Comment (calbera) [inline eth/core/precompile/method.go:95]:
```suggestion
			reflect.ValueOf(si),
```

**Review by itsdevbear — APPROVED** [review]

**Pull request merged**